proposals/NNNN-service-context.md

```diff
@@ -0,0 +1,177 @@
+|--|--|
+| **Package Name** | `swift-service-context` |
+| **Module Name** | `ServiceContextModule` |
+| **Proposed Maturity Level** | [Incubating](https://github.com/swift-server/sswg/blob/main/process/incubation.md#process-diagram) |
+| **License** | [Apache v2](https://github.com/apple/swift-distributed-tracing/blob/main/LICENSE.txt) |
+| **Dependencies** | - |
+
+## Introduction
+
+Service context is a type-safe dictionary that is keyed using types and intended to be used for contextual information propagation across asynchronous code.
+
+It declares a well-known Task Local key for the service context, and patterns how to add keys. 
+
+It is a core building block of [swift-distributed-tracing](https://github.com/apple/swift-distributed-tracing) which uses it to propagate trace information across swift concurrency tasks, though it can be used independently as well.
+
+## Overview
+
+Service context's API surface is relatively small and is all based around the service context type which is a type-safe dictionary.
+
+One can create a service context using the `.topLevel` factory method which intends to clarify that this should not be used UNLESS intending to start a fresh "top level" context that does not inherit any values from the current context:
+
+```swift
+var context = ServiceContext.topLevel
+```
+
+In most situations, developers should prefer to "pick up" the context from the current context:
+
+```swift
+var context: ServiceContext? = ServiceContext.current
+```
+
+which inspects the task-local values for the presence of a service context. It is by design that it is possible to determine if no context was set, or if an empty context is set. When intending to "carry existing, or create a new" context, the following pattern is used:
+
+```swift
+var context = ServiceContext.current ?? ServiceContext.topLevel
+```
+
+Once obtained, one can set values in by using defined keys, like this:
+
+```swift
+private enum FirstTestKey: ServiceContextKey {
+  typealias Value = Int
+}
+
+var context = ServiceContext.topLevel
+context[FirstTestKey.self] = 42
+```
+
+Keys declared as types conforming to `ServiceContextKey` allow for future extension where we might want to configure specific keys using additional behavior, like for example defining the `static let nameOverride` of a key. It is also important that a Key type may be private to whomever declares it, allowing only such library to set these values. 
+
+Service context is used primarily as a task-local value read and modified like this:
+
+```swift
+func exampleFunction() async -> Int {
+  guard let context = ServiceContext.current {
+    return 0
+  }
+  guard let value = context[FirstTestKey.self] {
+    return 0
+  }
+  print("test = \(value)") // test = 42
+  return value
+}
+
+// ----------------------------------------
+
+var context = ServiceContext.topLevel
+context[FirstTestKey.self] = 42
+
+let c = ServiceContext.withValue(context) {
+    await exampleFunction()
+}
+assert(c == 42)
```

Comment:
Seems like an odd example where you are returning the value to test it outside of the `withValue` function. Would it not be better to stick the assert inside the `withValue`.

Reply:
Ok sure, I'll rework a bit